# Third Hands-On Session
---
Exploring periodic lattices:

- Implementing and studying the properties of a FODO cell/lattice
- Computing closed optics solutions
- The concept of Tune and FODO cell phase advance
- Introducing Sector Bends
---

## 🐍 Python corner

Let's import standard packages and settings. We'll be using:
- `numpy` as `np`
- `matplotlib.pyplot` as `plt`
- functions `D`, `Q`, `transportParticles`, `getEquivalentElement`

In [1]:
# Import custom tracking functions and useful libraries
from tracking_library import *

numpy is installed, version: 2.3.1
scipy is installed, version: 1.15.3
matplotlib is installed, version: 3.10.0
-> Setup is OK! Have fun!


## ⚛️ Physics focus: From arbitrary transport to periodic transport

From the previous lecture we know that any beam line can be described by a transfer matrix M. In a periodic system (e.g. an accelerator ring), the beam passes through the same structure repeatedly, so after m turns:
\begin{equation}
X_m = M \cdot M \cdot ... \cdot M X_0 = M^m X_0
\end{equation}
where M is the transfer matrix of one full period, also called **the One-Turn Map (OTM)**. 

For the motion to be stable over many turns, the particle trajectory must remain bounded:
\begin{equation}
|X_m| < |\hat{X}| \qquad  \forall \quad X_0, m
\end{equation}
For a 2x2 real symplectic matrix (such as M), this stability condition translates into a simple requirement on the trace:
\begin{equation}
\frac{1}{2}|Tr(M)|\leq1
\end{equation}

**In the following, we will study particle motion and stability using one of the most common periodic structures: the FODO cell!**


## 📝 Exercise 3.1: The FODO cell

Let's start working with FODO cells. This is one of the most fundamental lattice topologies. In its simpler version, it consists on two thin quadrupoles of opposite focal length, f, spaced by two drifts of length $L$ (as in [Wolfgan's lecture](https://indico.cern.ch/event/1356988/contributions/5713241/)), i.e. the total length of the cell ($L_{FODO}$) is $2L$. 

<p align="center">
<img src=./_img_exercises/slide_fodo.png width="60%">
</p>

1. Define a **FODO beamline** that **starts and ends in the middle of a focusing quadrupole** (we could start the FODO from an arbitrary point inside it). For the moment, we can define arbitrarily the $f$ and $L$ parameters: we chose **2.5 m and 1 m**, respectively.

> 👀 **HINT 1**: A half-quadrupole has double the focal length than a full quadrupole.

> 👀 **HINT 2**: To better visualize the trajectory of the particle(s), always split the drifts in several shorter drifts (e.g. $10$ drifts, each $L/10$ long).
    
2. Track a particle that has an initial offset $x = 1$ mm and no angle $x' = -0.1$ mrad, and plot its position $x$ and angle $x'$ along the beam line.

> 👀 **HINT 3**: After having plotted $x$ trajectory on a plot, one can create a second vertical axis that shares the same horizontal axis with `plt.twinx()`

In [2]:
# code here your solution...

## 📝 Exercise 3.2: Multiple FODO cells - stability

Plot the position $x$ and angle $x'$ through $N=100$ equal FODO cells, play with different values of the focal length $f$ and explore whether you can make the oscillations grow.

> **Hint:** Look back at [Wolfgan's lecture](https://indico.cern.ch/event/1356988/contributions/5713241/):
>
> <p align="center">
> <img src=./_img_exercises/slide_stability.png width="60%">
> </p>
>
> It is shown that the stability of a FODO cell takes the form:
> \begin{equation}
    \frac{1}{2}|Tr(M)| = \left|\frac{L}{2f}\right| = \left|\frac{L_{FODO}}{4f}\right| \leq 1
> \end{equation}

In [3]:
# code here your solution...

## 📝 Exercise 3.3: Transport of sigma matrix in a FODO beamline & searching periodic solution

1. Transport a sigma matrix (e.g. $\beta$ = 3 [m], $\gamma$ = 0.5 [1/m], $\epsilon$ = 1) through a beamline made of 8 FODO cells (with, for example, $f=2.5$ and $L=1$ as before), and plot the evolution of the sigma matrix elements ($\sigma_{11}$, $\sigma_{22}$, $\sigma_{12}$) along the beamline. Do you see a periodic structure?

2. Try to vary the initial sigma matrix parameter by **try and error** that reproduces itself at the end of the beam line (8 FODO cells), meaning:

   \begin{equation}
      \Sigma_{0} = M_{\mathrm{OTM}} \Sigma_0 M_{\mathrm{OTM}}^T
   \end{equation}

 You will quickly realise that this is extremely difficult if not impossible. 

> 👀 **HINT 1**: (optional) you might want to use an **interactive plot** - see in the previous exercises how to make one

> 👀 **HINT 2**: **try!**, but **don't waist to much time with try and error** approaches (that's not your job as accelerator physicist!), and **move on** to the next exercise...

3. In a FODO cell the $\beta$ function reaches a maximum (or a minimum) in the middle of the quadrupoles, i.e. the $\alpha$ function is zero.
If one models the FODO cell starting from the middle of a quadrupole (as we do), then at least one parameter is fixed! 

In [4]:
# code here your solution...

## ⚛️ Physics focus: Periodic solution (matching)

Even when the motion is stable, the beam size (described by the sigma matrix $\Sigma$) may vary strongly from cell to cell or turn to turn. This can cause undesirable effects such as beam losses or emittance growth in the presence of nonlinearities.

In a circular accelerator, we usually want more than just bounded motion: we want the beam shape (phase space ellipse) to remain the same every turn. This special situation is called a **matched beam**, and it is defined by the condition:

   \begin{equation}
      \Sigma_{0} = M_{\mathrm{OTM}} \Sigma_0 M_{\mathrm{OTM}}^T
   \end{equation}

A beam satisfying this relation is in equilibrium with the lattice and remains stationary from turn to turn.


**Recalling** [Wolfgan's lecture](https://indico.cern.ch/event/1356988/contributions/5713241/): if we parameterize the beam with the Twiss parameters, the one-turn matrix of any stable beam line can be written as:

\begin{equation}
M_{\mathrm{OTM}} = 
    \left[
    \begin{array}{cc}
    \cos(\mu) + \alpha_0 \sin(\mu)  & \beta_0 \sin(\mu)\\
    - \gamma_0 \sin(\mu) & \cos(\mu) - \alpha_0 \sin(\mu)
    \end{array}
    \right]
    =
    \left[
    \begin{array}{cc}
    r_{11}  & r_{12} \\
    r_{21}  & r_{22}
    \end{array}
    \right]
\end{equation}

where $\alpha_0$, $\beta_0$, $\gamma_0$ are the initial (and final) **Twiss values** and $\mu$ is the phase advance (or **tune**, if we talk about a whole ring) of the associated beamline.


**From this representation one can directly extract the periodic Twiss parameters of any beamline, provided a stable solution exists, i.e. $\frac{1}{2}|\mathrm{Tr}(M)| \le 1$**

## 📝 Exercise 3.4: Matching to FODO lattice (periodic solution)

One can use the equation form [Wolfgan's lecture](https://indico.cern.ch/event/1356988/contributions/5713241/) to find the theoretical solution.

<p align="center">
<img src=./_img_exercises/slide_fodo_periodic.png width="60%">
</p>

Get the one turn matrix of the previous beamline using the `getEquivalentElement` function and extract the matching initial conditions. 

In [5]:
# code here your solution...

### 🔹 **NOTE** Those concepts are implemented in the function `twiss()` from our toolbox:

In [6]:
from tracking_library import twiss

help(twiss)

Help on function twiss in module tracking_library:

twiss(beamline)
    Compute Twiss parameters and tune for a periodic beamline.

    Parameters
    ----------
    beamline : list of dict
        Sequence of beamline elements.

    Returns
    -------
    tune : float
        Betatron tune (fraction of oscillation per turn).
    beta : float
        Beta function at entrance [m].
    alpha : float
        Alpha function at entrance.
    gamma : float
        Gamma function at entrance.



> 🔹 **NOTE**: We can apply this function to any beamline, e.g. a series of quadrupoles and drifts, e.g.:

In [7]:
# create a random beamline of many elements
beamline = 5 * (10 * D(0.5 / 10) + Q(1) + 10 * D(2 / 10) + Q(-0.8) + 10 * D(1.5 / 10) + Q(0.3) + 10 * D(1.5 / 10))

# compute the closed solution
(tune, beta, alpha, gamma) = twiss(beamline)

# assemble the sigma matrix
sigma_matrix = np.array([[beta, -alpha], [-alpha, gamma]])

# print out the values
print(f"The total phase advance (or tune) of the beamline is {tune}.")
print(f"The sigma matrix is:\n {sigma_matrix}")
print(f"which has determinant equal to: {np.linalg.det(sigma_matrix):.3f}.")

The total phase advance (or tune) of the beamline is 0.5999038489602736.
The sigma matrix is:
 [[16.98880317 14.57092107]
 [14.57092107 12.55601932]]
which has determinant equal to: 1.000.


❓**QUESTION: why the determinant of the obtained sigma matrix is equal to 1?**

## 📝 Exercise 3.5: Matching single FODO cell using `twiss`

Consider again a FODO cell, and choose $f$ and $L$ making sure it allows for having a periodic solution (i.e. $f > L_{FODO}/4$ - see again [Wolfgan's lecture](https://indico.cern.ch/event/1356988/contributions/5713241/)). 
This time, **model the FODO cell starting from the middle of a drift**, i.e. `D(L/2) + Q(f) + D + Q(-f) + D(L/2)`.

1. Compute the periodic $\Sigma_0$ matrix using the Twiss parameters obtain for the newly defined `twiss` function (assuming $\epsilon=1$).
2. Transport the periodic $\Sigma$ matrix along the FODO and convince yourself that the $\Sigma_s$ matrix at the end of the FODO is indeed equal to the one at the start, $\Sigma_0$.

In [8]:
# code here your solution...

## 📝 Exercise 3.6: Evolution of a matched simga matrix in multiple FODO cells
Write down the numerical values of initial beam matrix $\Sigma_0$, then build a beam line made of $15$ consecutive cells by changing the definition of the lattice and then, using $\Sigma_0$ with the noted-down numbers, prepare a plot of the beam sizes along the $15$ cells. 
Is this also periodic?

In [9]:
# code here your solution...

## 📝 Exercise 3.7: Trace (phase) space ellipse

Consider the previous FODO cell starting in the middle of the drift, and stable. 
Consider a single particle with some non-zero initial $x_0$ and $x_0'$ (e.g. $x_0 =.24$; $x' = 2.3$), and **plot the $x$, $x'$ coordinates at the end of each FODO cell** for a beamline made of 100 cells. 

> 👀 **HINT 1**: You should use both `getEquivalentElement` to obtain the single FODO cell element, and `transportParticles` function to track the particle over several of such a cell.

> 👀 **HINT 2**: use `plt.scatter(output['x'], output['px'])` to plot all phase-space coordinates.

### 🔹 NOTE: the Twiss parameters get a special meaning

<p align="center">
<img src=./_img_exercises/ellipse.png width="40%">
</p>

In [10]:
# code here your solution...

## 📝 Exercise 3.8: Betatron oscillations
Consider again the previous FODO cell, and a particle starting with non-zero coordinates. Plot the position of the particle vs the number of turns. 

1. What do you observe?
2. Can you infer the tune? 
3. How many turns (or FODO cells) are needed for the particle to complete one full oscillation?
4. What changes if you start from a different initial coordinate?

In [ ]:
# code here your solution...

## 📝 Exercise 3.9: Tuning FODO phase advance

Consider a single FODO cell of total length $L_{FODO}=2$ m. 

1. Can you find $f$ such that the FODO cell phase advance is $\mu = \pi/2$, i.e. 90 deg phase-advance?
2. Repeat this cell 100 times, and plot the trajectory, in phase space, of a particle with x=1 mm and xp=0 cell after cell: what do you observe?
3. (Optional) Show that the similar results are obtained irrespectively if you start your FODO from the middle of a quadrupole or of a drift.

> 👀 **HINT**: Look back at [Wolfgan's lecture](https://indico.cern.ch/event/1356988/contributions/5713241/):
>
> <p align="center">
> <img src=./_img_exercises/slide_fodo_phase.png width="60%">
> </p>
> 
> and recall that $\mathrm{Tr}(M) = 2 \cos(\mu)$. 
> With some math, one can prove that:
> 
>    \begin{equation}
       \sin(\mu/2) = \frac{L_{FODO}}{4f}
    \end{equation}

In [ ]:
# code here your solution...

## 📝 Exercise 3.10: Tuning a FODO cell with dipoles

Build a beam line made of 6 FODO cells with $L_{cell} = 4$ m and **Adjust the value of f** such that one cell phase advance is **60 degrees**.

1. Compute the closed optics using the `twiss()` function on a **single cell**, and plot the **beta function** along the whole beamline. Can you also compute the closed optics solution **applying twiss() to the whole beamline**? Does the solution differ? Why? What is the "tune" of the whole beamline?
2. Introduce **sector bends** ($L_{bend} = L_{DRIFT}/2 = 1$ m, and bending angle $\phi=10$ degrees) in the middle of all drifts of the FODO cells, without changing the total length of the FODO cell. How does the optics change?
3. Try to adjust the focal length of the quadrupoles to recover **60 degrees** phase advance **per cell**.

 > 👀 **HINT**: Our toolbox library implements also sector bends: `B(phi, L)` defined by bending angle (`phi`) and bend lenght (`L`).

In [2]:
# Import the sector bend model
from tracking_library import B

help(B)

Help on function B in module tracking_library:

B(phi, L)
    Thick sector bend with deflecting angle phi and length L (2x2 matrix).

    Parameters
    ----------
    phi : float
        Deflection angle [rad]
    L : float
        Length of the bend [m]

    Returns
    -------
    element : list of dict
        A list with a single dictionary containing the matrix and length.



In [ ]:
# code here your solution...

## Well done !!

=> **If you still have time, continue your learning with the following [notebook](./04_Dispersion.ipynb)**...
